Add Beersheba - the deconvolution city

Aretno opened the pull request:
First version of the deconvolution liquid city, named Beersheba (as it could have any other name). 

Flow of the city:

- Read esmeralda hits 
- Cut them based on their absolute SiPM Q and redistribute the energy of those cutted out proportionally to the rest of the track.
- Cut SiPMs that are isolated from the rest of the signal, that means that they do not have any neighbour with detected light. Again, redistribute the energy of the cut out hits.
- Apply deconvolution according to the configuration. At the moment applies deconvolution to each slice, in 2D.
- Assign the total energy of the peak to the deconvolved track. The assignment is weighted by the charge density of each deconvolved hit. 

Deconvolution features:
 - Signal can be interpolated or not before applying deconvolution.
 - Either uncorrected or corrected energy can be assigned.
 - Can use a one-step deconvolution (all EL and diffusion effects
 integrated within one PSF) or two-steps (first apply only-EL PSF
 deconvolution, then a gaussian-PSF deconvolution based on the drift
 of the hits).
 - Stopping criteria based on whatever is achieved first:
 maximum number of iterations or relative change between iterations threshold.

Things missing: tests and finish the implementation/check of 3D deconvolution. Aside from that everything works just fine.

PSF creation, a must for this, I think it should be outside IC, similar to KrCalib (or even inside?). The code for the PSF is mostly ready with some minor detail that is bugging me. In any case, PSF creation requires an especific data processing so this should be addressed in terms of the production chain.

Aretno committed:
Set of deconvolution used functions

Library with the different functions used in deconvolution:
- cut_and_redistribute_df basically cuts a dataframe on a given
condition and redistributes the values of the dataframe that are
passed.
- drop_isolated_sensors, as above, cuts a dataframe and redistributes
variables but based if a hit is considered isolated from the rest.
- deconvolutionInput prepares the data for deconvolution (puts it in a
grid and interpolates it if wanted, using interpolateSignal)
- find_nearest just finds the nearest array element to a given value,
useful to decide which PSF to use based on the Z position of the slice
- deconvolve basically applies the full deconvolution process,
including interpolation
- richardson_lucy is a modified version of the same function in
skimage, first the treatment of 0s has been changed as can lead to
nasty behaviour in the divisions, second the PSF mirroring has been
modified to make it n-dimensional, third an alternative stopping
criteria has been added to stop based on the relative change between
iterations and not on an absolute iteration number.

Aretno committed:
Add esmeralda hdst reader to be used in beersheba

Adds an esmeralda hdst reader needed for beersheba, only the Summary
information and the corrected Hits information is read and passed.

Aretno committed:
Beersheba city which applies track deconvolution

Beersheba reads esmeralda hits and applies deconvolution to
them. Afterwards, assign the cathode energy to the deconvolution
output and stores the deconvolved hits. Currently only tested for
slice by slice mode (2D deconvolution to each slice).

The PSF file that is used is a hdf5 containing a pandas dataframe
(GROUP: PSF, node: PSFs). Creation of the PSF shall go on another
commit not within IC but as a client, similar to KrCalib.

Features:
 - Signal can be interpolated or not before applying deconvolution.
 - Either uncorrected or corrected energy can be assigned.
 - Can use a one-step deconvolution (all EL and diffusion effects
 integrated within one PSF) or two-steps (first apply only-EL PSF
 deconvolution, then a gaussian-PSF deconvolution based on the drift
 of the hits).
 - Stopping criteria based on whatever is achieved first:
 maximum number of iterations or relative change between iterations threshold.

Aretno committed:
Configuration file for beersheba city

Aretno committed:
Tests for deconvolution functions

Test file for deconvolution tests

Handle division warnings

Aretno committed:
Added 3D deconvolution

Aretno committed:
Tests for beersheba city

Aretno committed:
Added cut mode

Option to cut output based on absolute or relative value

Aretno committed:
Small fixes

Fix gaussian definition and energy cut

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,330 @@
+                      deconv_mode     : str='joint',
+                      n_dim           : int=2,
+                      cut_type        : str='abs',
+                      interMethod     : str=None):
```
please don't use `camelCase` names for variables. We have adopted `lower_case_with_underscores` as a convention for variables and `CamelCase` for class names

gonzaponte replied:
Also, I would like some better naming for `iterationNumber`, `iterationThr` and `iterationGauss` (I suggest `n_iterations`, `iteration_tol` and `n_iterations_g` as they are kind of standard) (or a shorter version of `iterations` like `it` or `iter`)

gonzaponte replied:
`diffussion` should probably be something like an `XYZ` object, but I think we don't have one...

`energy_type` should be a `HitEnergy` (if it is not available, rebase)
create `AutoNameEnumBase` symbols for `deconv_mode`, `cut_type` and `interMethod`

Aretno replied:
I'll create the AutoNameEnumBase symbols, no problem at all but as you say, we do not have an XYZ object so I'll leave that as a tuple for the moment.

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,330 @@
+        if   cut_type == 'abs':
+            selDeconv   = deconvE > ecut
+        elif cut_type == 'rel':
+            selDeconv   = deconvE/deconvE.max() > ecut
```
```suggestion
            selDeconv   = deconvE / deconvE.max() > ecut
```

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,330 @@
+    deconvolution = deconvolve(iterationNumber, iterationThr, sampleWidth[:n_dim], bin_size[:n_dim], interMethod)
+    dimensions    = ['X', 'Y', 'Z'][:n_dim]
+
+    def create_deconvolution_df(hits, deconvE, pos):
```
I think you should make these functions standalone so you can test them independently

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,330 @@
+
+        return df
+
+    def distribute_energy(df, hdst):
```
This function modifies a column of an object and returns the object. The rule is in general that if you modify you don't return the object.

I think from the rest of the code it is safe to modify it because the scope of the df is coerced and it's not passed along outside the main function.

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,330 @@
+        return df
+
+    def distribute_energy(df, hdst):
+        df['E'] = df['E'] / df['E'].sum() * hdst[energy_type].sum()
```
```suggestion
        df['E'] = df.E / df.E.sum() * hdst.loc[:, energy_type].sum()
```

gonzaponte replied:
We have a situation.
This city uses a function that takes as a parameter the name of a column of a `DataFrame`, which is a `str`. That value comes to the city from the config file so it is also a `str`. However, the name of this column is limited to a few values from an `Enum` object so you find yourself doing something slightly weird:

```
class HitEnergy(AutoNameEnumBase):
    E  = auto()
    Ec = auto()
    Ep = auto()


@city
def beersheva(..., energy_type : str):
    e_type = getattr(HitEnergy, energy_type)
    def do_something(..., etype):
        column = dataframe[etype.value]
    do_something([...], e_type)
```

Notice how you take a `str`, get the corresponding `HitEnergy` object only to then access its `str` value again. The question here is whether there is a benefit to use the `Enum` object or not. On the one hand, you get a bit of documentation on what `energy_type` can be. On the other hand, it adds some lines to the code that may be distracting and not be very informative.

@jacg, I would like to know what's your take on this matter, if you have some time.

jacg replied:
@gonzaponte Thanks for the summary of the code (very useful), but I'm struggling to find `HitEnergy` anywhere in the real beersheba code!

IIUC, dataframes and the config files impose `str` which makes `HitEnergy` inconvenient. I would like to understand where `HitEnergy` is used throughout the code, before forming an opinion.

Approaches I can imagine, off the top of my head:

+ make the dataframe use the enum instead of strings, and translate it on the way in from the config file ... somewhere
+ accept and use a string in the city, but assert that the string matches one of the enum's members

In general, it would be preferable to use the Enum as much as possible (HitEnergy is so much more informative as a type than str, both for the human and the computer), but if this turns out to be convoluted, then it might be reasonable just to stick to strings. It's a compromise, but I'd need to know more details before forming an opinion.

gonzaponte replied:
> I'm struggling to find `HitEnergy` anywhere in the real beersheba code!

It had already been defined in [evm/event_model.py](https://github.com/nextic/IC/blob/master/invisible_cities/evm/event_model.py#L116) because it's also used for other pieces of code. The only use for beersheba is in this particular case. But this is actually another good point to use `HitEnergy`: it would be fishy to use `str` here but `HitEnergy` somewhere else when they actually mean the same.

> IIUC, dataframes and the config files impose `str` which makes `HitEnergy` inconvenient

I've just tried and dataframes can in principle take enums as columns both for initialization and column indexing. As for config files that's something that is currently true but that doesn't need to be: the functions that handle this could add all symbols to the context in which the config file is evaluated. For this, we would have to refactor the whole repo a little bit, though.

> I would like to understand where `HitEnergy` is used throughout the code, before forming an opinion.

If you mean in beersheva, there is only this case. For other uses, it is mostly used in [reco/paolina_functions.py](https://github.com/nextic/IC/blob/master/invisible_cities/reco/paolina_functions.py) and in [cities/esmeralda.py](https://github.com/nextic/IC/blob/master/invisible_cities/cities/esmeralda.py), which uses the former module.

> make the dataframe use the enum instead of strings, and translate it on the way in from the config file ... somewhere

I think this is a bit forced... We would have all column names to be strings but that one that is a `HitEnergy`.

> accept and use a string in the city, but assert that the string matches one of the enum's members

That was my second preferred option... right behind leaving it as is.

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,330 @@
+        if   deconv_mode == 'joint':
+            psf = psfs[(psfs.z == find_nearest(psfs.z,                 z)) &
+                       (psfs.x == find_nearest(psfs.x, df.Xpeak.unique())) &
+                       (psfs.y == find_nearest(psfs.y, df.Ypeak.unique())) ]
```
`psfs.loc[bla, :]`

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,330 @@
+        elif deconv_mode == 'separate':
+            psf_z0 = psfs[(psfs.z == find_nearest(psfs.z,                 0)) &
+                          (psfs.x == find_nearest(psfs.x, df.Xpeak.unique())) &
+                          (psfs.y == find_nearest(psfs.y, df.Ypeak.unique())) ]
```
`psfs.loc[bla, :]`

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,330 @@
+            psf_z0 = psfs[(psfs.z == find_nearest(psfs.z,                 0)) &
+                          (psfs.x == find_nearest(psfs.x, df.Xpeak.unique())) &
+                          (psfs.y == find_nearest(psfs.y, df.Ypeak.unique())) ]
+            deconvImage, pos = deconvolution(tuple(df[v].values for v in dimensions), df.Q.values, psf_z0)
```
```suggestion
            deconvImage, pos = deconvolution(tuple(df.loc[:, v].values for v in dimensions), df.Q.values, psf_z0)
```

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,330 @@
+
+    psfs          = pd.read_hdf(psf_fname)
+    deconvolution = deconvolve(iterationNumber, iterationThr, sampleWidth[:n_dim], bin_size[:n_dim], interMethod)
+    dimensions    = ['X', 'Y', 'Z'][:n_dim]
```
You should redefine all variables that need cropping here, like: `diffusion = diffusion[:n_dim]` and forget about doing it anymore in the rest of the code

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,330 @@
+                          (psfs.x == find_nearest(psfs.x, df.Xpeak.unique())) &
+                          (psfs.y == find_nearest(psfs.y, df.Ypeak.unique())) ]
+            deconvImage, pos = deconvolution(tuple(df[v].values for v in dimensions), df.Q.values, psf_z0)
+            g   = multivariate_normal(tuple(0. for _ in range(n_dim)), tuple((diff * sqrt(z/10))**2 for diff in diffusion[:n_dim])).pdf(list(zip(*tuple(psf_z0[f'{v.lower()}r'].values for v in dimensions))))
```
```suggestion
            dist     = multivariate_normal(np.zeros(n_dim), diff**2 * z / 10)
            cols     = tuple(f"{v.lower()}r" for v in dimensions)
            psf_cols = psf_z0.loc[:, cols]
            gaus     = dist.pdf(*psf_cols.values)
```
I haven't tried, but I think it's correct... Also, diff should be an array. Conversion should happen at the top.

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,330 @@
+                          (psfs.y == find_nearest(psfs.y, df.Ypeak.unique())) ]
+            deconvImage, pos = deconvolution(tuple(df[v].values for v in dimensions), df.Q.values, psf_z0)
+            g   = multivariate_normal(tuple(0. for _ in range(n_dim)), tuple((diff * sqrt(z/10))**2 for diff in diffusion[:n_dim])).pdf(list(zip(*tuple(psf_z0[f'{v.lower()}r'].values for v in dimensions))))
+            psf = g.reshape(tuple(psf_z0[f'{v.lower()}r'].nunique() for v in dimensions))
```
following previous comment...
```suggestion
            psf = gaus.reshape(psf_cols.nunique())
```

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,330 @@
+        deco_dst = []
+        for peak in sorted(set(df.npeak)):
+            deco_df = []
+            hits          = df[df.npeak == peak]
```
```
for peak, hits in df.groupby("npeak"):
    deco_df = []
```

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,330 @@
+            deco_df = []
+            hits          = df[df.npeak == peak]
+            hits._is_copy = False
+            hits['Q']     = hits.Q/hits.Q.sum()
```
Why this `_is_copy` thing? Is it to keep the changes of the Q normalization? If so, use `hits.loc[:, "Q"] = bla` after the applying previous comment.

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,330 @@
+            hits._is_copy = False
+            hits['Q']     = hits.Q/hits.Q.sum()
+            deconvolved_hits = (deconvolve_hits(hits, weighted_mean_and_std(hits.Z, hits.E)[0]) if n_dim == 3 else
+                                pd.concat([deconvolve_hits(hits[hits.Z == z], z) for z in hits.Z.unique()], ignore_index=True))
```
I think it is more readable in the form of:
```
if n_dim == 3: deconvolved_hits = one
else         : deconvolved_hits = the other
```

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,330 @@
+            hits['Q']     = hits.Q/hits.Q.sum()
+            deconvolved_hits = (deconvolve_hits(hits, weighted_mean_and_std(hits.Z, hits.E)[0]) if n_dim == 3 else
+                                pd.concat([deconvolve_hits(hits[hits.Z == z], z) for z in hits.Z.unique()], ignore_index=True))
+            deco_dst.append(distribute_energy(deconvolved_hits, hits))
```
If you apply the comment on this function be careful to take the call out of the append or you will be appending `None`s!

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,330 @@
+    return write_deconv
+
+
+def summary_writer(h5out, compression='ZLIB4', group_name='SUMMARY', table_name='Events', descriptive_string='Event summary information', str_col_length=32):
```
why is `str_col_length` an argument?

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,330 @@
+    SUMMARY : Table with the summary from Esmeralda.
+"""
+
+    deconv_params_   = {k : v for k, v in deconv_params.items() if k not in ['q_cut', 'drop_dist']}
```
Mmm, I think the name is not very appropriate because can be very easily confused with the original... I would recommend
```
q_cut     = deconv_params.pop("q_cut")
drop_dist = deconv_params.pop("drop_dist")
# here deconv_params no longer contains those two entries 
```

gonzaponte commented on invisible_cities/cities/beersheba_test.py:
```diff
@@ -0,0 +1,161 @@
+from .. core.testing_utils     import assert_tables_equality
+
+
+@mark.serial
```
why serial? does it depend on another test? If so, maybe it's better to save the output of that test as a permanent test file to input here

gonzaponte commented on invisible_cities/cities/beersheba_test.py:
```diff
@@ -0,0 +1,161 @@
+                     deconv_params     = dict(
+                                         q_cut           =         10,
+                                         drop_dist       = [10., 10.],
+                                         psf_fname       = f'{ICDIR}database/test_data/PSF_dst_sum_collapsed.h5',
```
define this path as a fixture in `conftest.py` and include it as an argument to the test

gonzaponte commented on invisible_cities/cities/beersheba_test.py:
```diff
@@ -0,0 +1,161 @@
+                assert_tables_equality(got, expected)
+
+
+def test_beersheba_empty_input_file(ICDIR, config_tmpdir, ICDATADIR):
```
what is this test supposed to check?

Aretno replied:
This is just a safety check that the city does not crash when trying to run over empty files. This has been added to some cities to make sure that the production daemons are not stopped when this happens.

gonzaponte commented on invisible_cities/cities/components.py:
```diff
@@ -280,6 +280,35 @@ def pmap_from_files(paths):
             # it needs to be given the WHOLE TABLE (rather than a
             # single event) at a time.
 
+def hdst_from_files(paths: List[str]) -> Iterator[Dict[str,Union[pd.DataFrame, MCInfo, int, float]]]:
```
isn't there already a reader for this? For sure there is one with hdst and kdst.

gonzaponte replied:
Esmeralda has been approved, so I guess you will need to discuss with @mmkekic how to organize the readers since they are very similar. Also I think some things (such as persistent data representation) have changed.

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -0,0 +1,288 @@
+from typing  import Callable
+
+from scipy        import interpolate
+from scipy.signal import fftconvolve, convolve
```
split import into two

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -0,0 +1,288 @@
+        pass_df._is_copy = False
+        with np.errstate(divide='ignore'):
+            for redist_variable in variables:
+                pass_df[redist_variable] = pass_df[redist_variable] * np.divide(df[redist_variable].sum(), pass_df[redist_variable].sum())
```
I think the following might work:

```
columns  = pass_df.loc[:, variables]
columns *= np.divide(df.loc[:, variables].sum(), columns)
```
which will avoid the `_is_copy` thing if it does what I think it does

gonzaponte replied:
```
    def cut_and_redistribute(df : pd.DataFrame) -> pd.DataFrame:
        pass_df = df.query(cut_condition)
        if not len(pass_df): return pass_df

        with np.errstate(divide='ignore'):
            columns  =      pass_df.loc[:, variables]
            columns *= np.divide(df.loc[:, variables].sum(), columns.sum())
        return pass_df
```

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -0,0 +1,288 @@
+
+    return cut_and_redistribute
+
+def drop_isolated_sensors(distance  : List[float]=[10., 10.],
```
no need for the decimal points

jahernando replied:
This function could be of general use. I will propose to change the name. Something like *select_hits_and_redestribute_signal* and I guess in the variables will be something like "E" or "Ec", right? Maybe given a default could be useful.

mmkekic replied:
As I understand this function is doing the same as reco/hits_functions but using directly dataframes instead of HitCollection type. Maybe we can then move this function to reco/hits_functions and eventually rewrite the old code (ie esmeralda) to use this one since it is faster than converting to our HitCollection data type?

gonzaponte replied:
We have decided to leave that for another PR

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -0,0 +1,288 @@
+        y       = df.Y.values
+        mask_xy = [False if np.ma.masked_equal(np.sqrt((x - xi)**2 + (y - yi)**2), 0.0, copy=False).min() > dist
+                         else True
+                         for xi, yi in zip(x, y)]
```
```
dr2     = (x - x.T)**2 + (y - y.T)**2 # compute all square distances
closest = np.apply_along_axis(lambda d: d[d > 0].min(), 1, dr2) # find closest that it's not itself
mask_xy = closest <= dist**2 # take those with at least one neighbour
```
I don't know if this could be more clear like this

jahernando replied:
again here, this function could be of general use to filter 'hits' that have not close neighbors. Maybe a more general name *select_hits_with_neighbors_in_distance*. 

I do not see why we pass (x, y)-distance when, if I understand right, there is one the radius used.

Aretno replied:
Could be only a single distance but I feel it's more intuitive to give the pitch en each dimension. I somehow suspect that if we just write the distance the user may just assume that to keep the first ring is enough with 1 pitch distance instead of the 1.41 * pitch but I do not have a strong opinion here, it can be easily changed.

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -0,0 +1,288 @@
+        pass_df._is_copy = False
+        with np.errstate(divide='ignore'):
+            for redist_variable in variables:
+                pass_df[f'{redist_variable}'] = pass_df[redist_variable] * np.divide(df[redist_variable].sum(), pass_df[redist_variable].sum())
```
same tricks as before

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -0,0 +1,288 @@
+                           ) -> Tuple[np.ndarray, Tuple[np.ndarray, ...]]:
+        ranges  = [[data[i].min() - 1.5 * sw, data[i].max() + 1.5 * sw] for i, sw   in enumerate(sampleWidth)]
+        allbins = [int(np.diff(rang)/sampleWidth[i])                    for i, rang in enumerate(     ranges)]
+        nbin    = [int(np.diff(rang)/bin_size[i])                       for i, rang in enumerate(     ranges)]
```
don't you need to `ceil` it?

Aretno replied:
This is kind of delicate, so... in reality you are right but this is a 'patchy' solution to some problem related of the uneven distribution of the Z associated to the slices. So... since slices are not separated equally in 2 us (at least the stored Z value) but distributed around 2 (but actually usually lower, as we talked at some point) I had several 0s when trying to interpolate. I decided on int so it would make bins lower than the rebinning in order to avoid 0s in the histogram as much as possible. 

However, I may try to just use uneven binning since that is a possibility, that should also work and remove all kind of binning problems.

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -0,0 +1,288 @@
+        interPoints = tuple      (interP.flatten() for interP in interPoints)
+
+        if interMethod is not None:
+            Hs, interPoints    = interpolateSignal(Hs        , interPoints    , edges    , nbin    , interMethod)
```
why the spaces?

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -0,0 +1,288 @@
+
+    H1 = interpolate.griddata(interPoints, Hs.flatten(), newPoints, method=interMethod)
+    H1 = np.nan_to_num       (H1.reshape(nbin))
+    H1 = np.where            (H1<0, 0, H1)
```
```suggestion
    H1 = np.clip             (H1, 0, None)
```

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -0,0 +1,288 @@
+                  ) -> Tuple[np.ndarray, Tuple[np.ndarray, ...]]:
+
+        interSignal, interPos = deconvInput(data, weight)
+        psf_deco      = psf.factor.values.reshape(tuple(psf[var].nunique() for var in varName[:len(data)]))
```
```suggestion
        columns       = varName[:len(data)]
        psf_deco      = psf.factor.values.reshape(psf.loc[:, columns]].nunique().values)
```

gonzaponte commented on invisible_cities/reco/deconv_functions_test.py:
```diff
@@ -0,0 +1,174 @@
+
+    assert len(df_cut) == len(df_cut_manual)
+
+    if (len(df_cut)) != 0:
```
```suggestion
    if len(df_cut):
```

gonzaponte commented on invisible_cities/reco/deconv_functions_test.py:
```diff
@@ -0,0 +1,174 @@
+    drop_function = drop_isolated_sensors(dist, ['E'])
+    df_cut        = drop_function(df)
+
+    if len(df_cut > 0):
```
Did you mean `len(df_cut) > 0`?

gonzaponte commented on invisible_cities/reco/deconv_functions_test.py:
```diff
@@ -0,0 +1,174 @@
+
+    for row in df_cut.itertuples(index=False):
+        n_neighbours = len(df_cut[in_range(df_cut.X, row.X - dist[0], row.X + dist[0] + 1e-8) & in_range(df_cut.Y, row.Y - dist[1], row.Y + dist[1] + 1e-8)])
+        assert n_neighbours > 1
```
I don't like that `+1e-8`. I think this is a needed feature in `in_range`: add arguments to decide which comparison to use on each side. For now maybe you can increase the distance to `10.1`?

gonzaponte replied:
Also split the line in two

gonzaponte commented on invisible_cities/reco/deconv_functions_test.py:
```diff
@@ -0,0 +1,174 @@
+
+    assert np.allclose(ref_interpolation, np.around(out_inter[0].flatten(), decimals=3))
+    assert np.allclose(shift_to_bin_centers(np.linspace(-0.05, 1.05, n_interpolation + 1)), sorted(set(out_inter[1][0])))
+    assert np.allclose(shift_to_bin_centers(np.linspace(-0.05, 1.05, n_interpolation + 1)), sorted(set(out_inter[1][1])))
```
give meaningful name to the `out_inter` items and to the `np.linspace` outcomes for a better readability of the test

gonzaponte commented on invisible_cities/reco/deconv_functions_test.py:
```diff
@@ -0,0 +1,174 @@
+    xx, yy = xx.flatten(), yy.flatten()
+
+    psf           = {}
+    psf['factor'] = multivariate_normal([0., 0.], [1.027 * np.sqrt(z/10)] * 2).pdf(list(zip(xx, yy)))
```
aren't you missing a `**2`?

gonzaponte commented on invisible_cities/reco/deconv_functions_test.py:
```diff
@@ -0,0 +1,174 @@
+    xx, yy = np.meshgrid(x, y)
+    xx, yy = xx.flatten(), yy.flatten()
+    psf = {}
+    psf['factor'] = multivariate_normal([0., 0.], [1.027 * np.sqrt(z/10)] * 2).pdf(list(zip(xx, yy)))
```
`**2` missing?

Aretno replied:
Well, 'physics-ally' speaking yeah but in the end that only broadens or not the PSF. To test the functions is enough that the PSF used to generate the referencefe file and the one used in the function is the same, which is the case.

gonzaponte commented on invisible_cities/reco/deconv_functions_test.py:
```diff
@@ -0,0 +1,174 @@
+    psf['zr']     = [z] * len(xx)
+    psf   = pd.DataFrame(psf)
+
+    deco = richardson_lucy(inter[0], psf.factor.values.reshape(psf.xr.nunique(), psf.yr.nunique()).T, 15, 0.0001)
```
maybe use `iterations = 50, iter_thr = 1e-4` for a better understanding of the function call

gonzaponte reviewed (commented):
Please review the naming of the variables and use `lower_case_with_underscores` instead of `camelCase` to blend into the general style of IC.

This was not a complete review, only what I've seen reading the code quickly.

Aretno committed:
Address PR comments

Aretno committed:
Address PR comments

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,355 @@
+from .. types.ic_types        import AutoNameEnumBase
+
+
+class CutType          (AutoNameEnumBase):
```
wow, not even me would have done that alignment! Good job! :laughing:

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,355 @@
+                      iteration_tol   : float,
+                      sample_width    : List[float],
+                      bin_size        : List[float],
+                      diffusion       : Tuple[float]=(1., 1., 0.3),
```
couldn't/shouldn't these numbers come from the database?

Aretno replied:
I think those values are not on the database as of now, right?

gonzaponte replied:
I thought they were there, but apparently not. Maybe they are stored in the map files? I don't know. Anyway, ignore it if it doesn't make sense,

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,355 @@
+    Parameters
+    ----------
+    psf_fname       : Point-spread function.
+    e_cut            : Cut in relative value to the max voxel over the deconvolution output.
```
slight missalignment

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,355 @@
+    dimensions    = np.array(['X', 'Y', 'Z'][:n_dim])
+    sample_width  = np.array(sample_width   [:n_dim])
+    bin_size      = np.array(bin_size       [:n_dim])
+    diffusion     = np.array(diffusion      [:n_dim])
```
I am a bit concern about the code taking `sample_width`,  `bin_size` and  `diffusion` of the wrong shape without complaining.

If `n_dim == 2`, those variables can have whatever dimension > 2 and everything will still work, which may be puzzling. In the same way, it those variables have dimension 1, this piece of code will not crash and the error (if ever raised) will be at a random point, which can also be confusing. Maybe it's better to assert that they have the right dimensions? It would be interesting to have other people's opinions here, as I am not fully convinced.

Also, `np.asarray` will avoid copies whenever possible, which is a very subtle detail, but it informs the reader that you want it to be an array, not necessarily to make a copy of it (which `np.array` does).

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,355 @@
+                      deconv_mode     : str='joint',
+                      n_dim           : int=2,
+                      cut_type        : str='abs',
+                      inter_method    : str='cubic'):
```
These should be `Enum` objects directly, not `str` that you use to access the object. Part of the idea of enums is that they are bounded to a few options and that they are self-documented. If one of your arguments is type-hinted as (for instance) `DeconvolutionMode` you automatically know what needs to go there. Also, this way the conversion from `str` to `Enum` happens only once, and at the top level of the program.

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,355 @@
+
+    e_type        = HitEnergy(energy_type)
+    cut_mode      = CutType(cut_type)
+    deco_mode     = DeconvolutionMode(deconv_mode)
```
Related to the comment about `Enums` before.

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,355 @@
+        elif deco_mode is DeconvolutionMode.separate:
+            psf_z0 = psfs.loc[(psfs.z == find_nearest(psfs.z,                 0)) &
+                              (psfs.x == find_nearest(psfs.x, df.Xpeak.unique())) &
+                              (psfs.y == find_nearest(psfs.y, df.Ypeak.unique()))  , :]
```
I don't want to be too pedantic, but I would prefer to have `x = df.Xpeak.unique()` (and `y`) outside the if statement because they are common to both and it would reduce the noise in those already complicated expressions.

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,355 @@
+            psf_z0 = psfs.loc[(psfs.z == find_nearest(psfs.z,                 0)) &
+                              (psfs.x == find_nearest(psfs.x, df.Xpeak.unique())) &
+                              (psfs.y == find_nearest(psfs.y, df.Ypeak.unique()))  , :]
+            deconv_image, pos = deconvolution(tuple(df.loc[:, v].values for v in dimensions), df.Q.values, psf_z0)
```
`df.loc[:, dimensions].values.T`

Also, this line could be outside the if statement, as it is the same for both cases. Even more, the whole thing is the same with the exception that you use the pdf for `z == some_z` or for `z == 0`. I suggest you define a new variable that is either ' = z` or ` = 0' .

Finally I would also add an `else` statement at the end for an invalid `deco_mode`

Aretno replied:
I don't think the `else` statement is need as the validity of the deco_mode is tested once you call the Enum object, right?

Aside from that, I'm ok with the rest.

gonzaponte replied:
> the validity of the deco_mode is tested once you call the Enum object, right?

Not sure what you are you saying here. If you mean _"if it is not this value one, it must be the other one"_ then it is not the case. This is a common situation that leads to bugs. For instance, you may mistake this argument with another one and it wouldn't fail cause there is not an exhaustive check of the validity of the arguments. `Enum`s are used precisely to narrow down the allowed values of a given parameter to just a few options. If the user sends a wrong argument it should be detected.

Aretno replied:
What I'm trying to say is that, when you create a DeconvolutionMode() object, if the argument passed is not one of the available modes, then it will fail and you won't even get to the inside of that function. 

The argument is a string which is converted at the beginning to a DeconvolutionMode object. The DeconvolutionMode object will fail (rightfully so) at the very beginning if the argument is not part of the `Enum` possibilites. So if you give a non-string argument to the city, it will fail because the DeconvolutionMode object won't swallow it, if it's a string, then it will fail if the string is not in the considered modes in the `Enum`. 

Or is there something else that I'm missing?

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,355 @@
+    def apply_deconvolution(df):
+        deco_dst = []
+        for peak, hits in df.groupby("npeak"):
+            hits.loc[:, "Q"] = hits.Q/hits.Q.sum()
```
IIRC, you need to compute normalized charges many times in your code. If that is the case you may want to do this to the full dataframe:
```
df.loc[:, "Qnorm"] = df.groupby(["event", "npeak"]).Q.transform(lambda q: q / q.sum())
```
and then use this new column everywhere.
At the same time you don't modify the original column, which is always dangerous.

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,355 @@
+            if n_dim == 3:
+                deconvolved_hits =            deconvolve_hits(hits, weighted_mean_and_std(hits.Z, hits.E)[0])
+            else :
+                deconvolved_hits = pd.concat([deconvolve_hits(hits.loc[hits.Z == z, :], z) for z in hits.Z.unique()], ignore_index=True)
```
This line would be more optimal doing something like `hits.groupby("Z").aggregate(deconvolve_hits)`, but it would need an intermediate function to work. Maybe `pd.concat([deconvolve_hits(df_z, z) for z, df_z in hits.groupby("Z")])`

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,355 @@
+            distribute_energy(deconvolved_hits, hits, e_type)
+            deco_dst.append(deconvolved_hits)
+
+        return pd.concat(deco_dst, ignore_index=True)
```
I think the whole function could be re-thought. Because what you do here is split-apply-combine, for which pandas provides better ways...
In this case, I think I would define the inner code that does `if_ndim == 3 ... else ...` as a function and then do `df.groupby(["event", "npeak"].transform(apply_deconvolution)` (maybe `apply` instead of `transform`)

Aretno replied:
After rewriting the function, the `apply` approach takes almost twice the time running (checked with cProfile) than the manual groupby-split, therefore I'm reluctant to do the `apply` approach.

jacg replied:
I recommend that you commit this approach, then follow it with a commit that reverts it. In the commit message titles provide clear indication that this was tried and rejected for performance reasons. In the body of the messages provide the profiling / benchmarking results. Keep it in the branch history after merging, even though the code doesn't go into production.

You want the log to document that this was tried, what *exactly* was tried, in a retrievable way. Like that:

1. The fact that it was tried and rejected is documented.
2. If we ever suspect the implementation that was tried, or don't trust the timings, it will be easy to reproduce.

Aretno replied:
To add a bit more info on this, apply, at least on the current pandas version on IC, is called twice over the first group (https://stackoverflow.com/questions/21390035/pandas-groupby-apply-method-duplicates-first-group). This I imagine results in calling the deconvolution process an additional time and adding a humongous processing time.

However, newer versions of pandas do not longer show this behaviour (https://github.com/pandas-dev/pandas/pull/24748). I tested the code using this pandas version and the running time goes back to 'normal'. Therefore, either we update pandas (which could be delicate as I'm unsure if it would have any collateral on other functions) or we use the initial manual approach for splitting.

Aretno replied:
Oh, I did not see your answer, @jacg. Will do as you propose!

gonzaponte replied:
> either we update pandas (which could be delicate as I'm unsure if it would have any collateral on other functions) 

We need to keep up to date with the packages. Otherwise, everything will get more and more complicated to update in the future. No need to do it *right now*, but maybe we should implement a regular protocol to review packages and update them once a year or so. We should start by opening an issue and decide which packages to update.


As for the code, I am a bit annoyed that we cannot use `pandas`' built-ins, but maybe it's not so important...

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,355 @@
+
+
+def distribute_energy(df, hdst, e_type):
+    df['E'] = df.E / df.E.sum() * hdst.loc[:, e_type.value].sum()
```
Do you want to overwrite the energy or to create a new column?
Use `df.loc[:, "E"]` for the left side

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,355 @@
+    """
+    def write_summary(df):
+        return _store_pandas_as_tables(h5out = h5out, df = df, compression = compression, group_name = group_name, table_name = table_name, descriptive_string = descriptive_string)
+    return write_summary
```
I would think the arguments `group_name`, `table_name` and `descriptive_string` are not really modifiable in any practical case, right? In that case I think I would rather hardcode them in the inner function.

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,355 @@
+"""
+    cut_sensors       = fl.map(cut_over_Q   (deconv_params.pop("q_cut"), ['E', 'Ec']),
+                               args = 'hdst',
+                               out  = 'hdst_cut')
```
Many of the operations performed are overwriting the input dataframe, isn't it? In that case `hdst` and `hdst_cut` would hold the same dtaframe so I think it would be fairer to use `item = 'hdst'`, which is equivalent of `args=whatever, out=whatever`

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,355 @@
+
+    drop_sensors      = fl.map(drop_isolated(deconv_params.pop("drop_dist"), ['E', 'Ec']),
+                               args = 'hdst_cut',
+                               out  = 'hdst_drop')
```
same, if `drop_isolated` modifies the input in place I think `item` it's more correct.

gonzaponte commented on invisible_cities/cities/beersheba_test.py:
```diff
@@ -0,0 +1,185 @@
+    true_in  = os.path.join(ICDATADIR, "exact_Kr_deconvolution_with_MC.h5")
+    true_dst = dio.load_dst(true_in, 'DECO', 'Events')
+    ecut = 1e-2
+    new_dst  = pd.concat([create_deconvolution_df(t, t.E.values, (t.X.values, t.Y.values, t.Z.values),
+                                                  CutType.abs, ecut, 3) for _, t in true_dst.groupby('event')])
```
Do you need to do `.values` for it to work? Seems a bit annoying

Aretno replied:
Yup, otherwise the selection inside `create_deconvolution_df` won't work properly and return nan values here:

```    
df['X']     = pos[0][sel_deconv]
df['Y']     = pos[1][sel_deconv]
```

as `df` indexes do not  match `pos[sel_deconv]` (with pos equivalent to (t.X, t.Y)). Dunno if there is an easy workaround but the function is thought to work with arrays as an input.

gonzaponte replied:
You could use `np.asarray` within `create_deconvolution_df` whenever you need to. This will move the ugliness to the guts of the function. Not sure if it is worth it, though...

gonzaponte commented on invisible_cities/cities/beersheba_test.py:
```diff
@@ -0,0 +1,185 @@
+def test_create_deconvolution_df(ICDATADIR):
+    true_in  = os.path.join(ICDATADIR, "exact_Kr_deconvolution_with_MC.h5")
+    true_dst = dio.load_dst(true_in, 'DECO', 'Events')
+    ecut = 1e-2
```
align

gonzaponte commented on invisible_cities/cities/beersheba_test.py:
```diff
@@ -0,0 +1,185 @@
+
+
+def test_distribute_energy(ICDATADIR):
+    true_in  = os.path.join(ICDATADIR, "exact_Kr_deconvolution_with_MC.h5")
```
align

gonzaponte commented on invisible_cities/cities/beersheba_test.py:
```diff
@@ -0,0 +1,185 @@
+    distribute_energy(true_dst2, true_dst1, HitEnergy.E)
+
+    assert np.allclose(true_dst2.E.values/true_dst2.E.sum(), true_dst3.E.values/true_dst3.E.sum())
+    assert true_dst1.E.sum() == true_dst2.E.sum()
```
If these are floats use `np.isclose` or `approx` (from pytest)

gonzaponte commented on invisible_cities/cities/beersheba_test.py:
```diff
@@ -0,0 +1,185 @@
+        assert "DECO/Events"             in h5out.root
+        assert "Run"                     in h5out.root
+        assert "Run/events"              in h5out.root
+        assert "Run/runInfo"             in h5out.root
```
remove unnecessary spaces

gonzaponte commented on invisible_cities/cities/beersheba_test.py:
```diff
@@ -0,0 +1,185 @@
+                                         deconv_mode     =    'joint',
+                                         n_dim           =          2,
+                                         cut_type        =      'rel',
+                                         inter_method    = 'cubic')))
```
review alignment. Also, I think the parameters of these tests should be given by a fixture, which should probably also yield the filename.

gonzaponte commented on invisible_cities/cities/beersheba_test.py:
```diff
@@ -0,0 +1,185 @@
+                                         deconv_mode     =         'joint',
+                                         n_dim           =               3,
+                                         cut_type        =           'rel',
+                                         inter_method    = 'linear')))
```
same comment about the parameters and the true output file

gonzaponte commented on invisible_cities/cities/beersheba_test.py:
```diff
@@ -0,0 +1,185 @@
+                                         n_dim           =          2,
+                                         cut_type        =      'abs',
+                                         inter_method    = 'cubic')))
+
```
same

gonzaponte commented on invisible_cities/cities/beersheba_test.py:
```diff
@@ -0,0 +1,185 @@
+                assert_tables_equality(got, expected)
+
+
+def test_beersheba_empty_input_file(ICDIR, ICDATADIR, PSFDIR, config_tmpdir):
```
Mmm, what's the point of this test? I think that checking that it doesn't crash for empty files has nothing to do with a specific city. If the input file is empty the reader will not generate any data and the whole dataflow thing will stop immediately. I am not saying that you should remove it, but maybe it's more a test for the reader rather than the city

Aretno replied:
I think this was basically because in some cities (don't remember which), before, the city crashed when trying to run on an empty file and it caused the data processing daemon in Canfranc to stop working due to that.    So this basically checks that this does not happen when running the city.

gonzaponte commented on invisible_cities/cities/components.py:
```diff
@@ -283,6 +283,35 @@ def pmap_from_files(paths):
+            for evtinfo in event_info:
+                event_number, timestamp = evtinfo.fetch_all_fields()
+                yield dict(hdst    = hdst_df   .loc[hdst_df   .event==event_number],
+                           summary = summary_df.loc[summary_df.event==event_number],
```
maybe it's better to create a `groupby` object outside the loop and then just do `get_group` for each iteration. But I don't know if there will be problems if any of the dataframes have an event number missing...

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -0,0 +1,327 @@
+            mask_xy = closest <= dist # take those with at least one neighbour
+            pass_df = df.loc[mask_xy, :]
+        else:
+            return df.loc[[False]*len(df)]
```
Assuming that the else statement means "return empty dataframe", I suggest the following implementation.
```
if not np.any(dr2 > 0):
    return df.iloc[:0] # Empty dataframe
```
And there it no need to have an `else` statement. The rest of the code can be with the normal indentation

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -0,0 +1,327 @@
+                           ) -> Tuple[np.ndarray, Tuple[np.ndarray, ...]]:
+
+        ranges  = [[data[i].min() - 1.5 * sw, data[i].max() + 1.5 * sw]             for i, sw   in enumerate(sample_width)]
+        nbin    = [np.ceil(np.diff(rang)/bin_size[i]).astype('int')[0]              for i, rang in enumerate(  ranges    )]
```
Since the index `i` is used only to access an element of another sequence I would prefer:
```
ranges = [[coord.min() - 1.5 * sw, coord.max() + 1.5 * sw]             for coord, sw  in zip(data, sample_width)]
nbin   = [int(np.diff(rang)[0]/bs)                                     for bs, rang   in zip(bin_size,  ranges )]
```
(also notice the simpler expression for the second line)

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -0,0 +1,327 @@
+        nbin    = [np.ceil(np.diff(rang)/bin_size[i]).astype('int')[0]              for i, rang in enumerate(  ranges    )]
+
+        if inter_method is not InterpolationMethod.none:
+            allbins = [np.linspace(*rang, np.ceil(np.diff(rang)/sample_width[i])+1) for i, rang in enumerate(  ranges[:2])]
```
```suggestion
            allbins = [np.linspace(*rang, int(np.diff(rang)/sw) + 1) for rang, sw in zip(  ranges[:2], sample_width)]
```

gonzaponte replied:
Check alignment after implementing the change

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -0,0 +1,327 @@
+                bins_z[ :2 ] = data[2].min() - np.array([1.5, 0.5]) * mean_diff
+                bins_z[-2: ] = data[2].max() + np.array([0.5, 1.5]) * mean_diff
+                allbins.append(bins_z)
+
```
Maybe add some comment about what happens here and why

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -0,0 +1,327 @@
+    H1         : Interpolated distribution weights.
+    new_points : Interpolated coordinates.
+    """
+    new_points   = np.meshgrid(*(shift_to_bin_centers(np.linspace(np.min(edge), np.max(edge), nbin[i]+1)) for i, edge in enumerate(edges)), indexing='ij')
```
```suggestion
    new_points   = np.meshgrid(*(shift_to_bin_centers(np.linspace(np.min(edge), np.max(edge), n + 1)) for n, edge in zip(nbin, edges)), indexing='ij')
```

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -0,0 +1,327 @@
+    return deconvolve
+
+def richardson_lucy(image, psf, iterations=50, iter_thr=0.):
+    """Richardson-Lucy deconvolution (modification from scikit-image package).
```
Specify the main differences with the scikit-image method

gonzaponte commented on invisible_cities/reco/deconv_functions_test.py:
```diff
@@ -0,0 +1,179 @@
+characters = string.ascii_letters
+
+@composite
+def dst(draw, dimension=[3,10]):
```
hypothesis can generate dataframes, are those not good for your purposes?

gonzaponte commented on invisible_cities/reco/deconv_functions_test.py:
```diff
@@ -0,0 +1,179 @@
+@given(dst())
+def test_cut_and_redistribute_df(df):
+    cut_var       = random.choice (df.columns)
+    redist_var    = random.choices(df.columns, k=3)
```
random stuff in a hypothesis-based test is not accepted. If the test were to fail, hypothesis would try reproduce it, but because it's random it will not be able to. Some options are:
- Fix the columns on which you perform the operations. After all, these are all hypothesis-generated numbers, so you can't have any control over them and it doesn't matter which columns you pick.
- Accept also an integer (or many) in the adequate range(s) to select the columns
- Doing the same, but creating a new composite that does that
- Use a random generator from hypothesis (not the most elegant in my opinion)

gonzaponte commented on invisible_cities/reco/deconv_functions_test.py:
```diff
@@ -0,0 +1,179 @@
+    df_cut        = cut_function(df)
+    df_cut_manual = df.loc[df[cut_var].values > cut_val, :]
+    for var in redist_var:
+        df_cut_manual.loc[:, var] = df_cut_manual.loc[:, var] * df.loc[:, var].sum() /  df_cut_manual.loc[:, var].sum()
```
I think you can remove the loop if you just replace `var` with `redist_var` in the second line.

gonzaponte commented on invisible_cities/reco/deconv_functions_test.py:
```diff
@@ -0,0 +1,179 @@
+def test_drop_isolated_sensors():
+    size          = 20
+    dist          = [10.1, 10.1]
+    x, y          = random.choices(np.linspace(-200, 200, 41), k=size), random.choices(np.linspace(-200, 200, 41), k=size)
```
maybe this line is unnecessarily long and it can be split

gonzaponte commented on invisible_cities/reco/deconv_functions_test.py:
```diff
@@ -0,0 +1,179 @@
+
+    h = hdst[(hdst.event == 3021916) & (hdst.npeak == 0)]
+    h = h.groupby(['X', 'Y']).Q.sum().reset_index()
+    h = h[h.Q>40]
```
```suggestion
    h = h[h.Q > 40]
```

gonzaponte commented on invisible_cities/reco/deconv_functions_test.py:
```diff
@@ -0,0 +1,179 @@
+    h = hdst[(hdst.event == 3021916) & (hdst.npeak == 0)]
+    z = h.Z.mean()
+    h = h.groupby(['X', 'Y']).Q.sum().reset_index()
+    h = h[h.Q>40]
```
```suggestion
    h = h[h.Q > 40]
```

gonzaponte reviewed (commented):
Second round. We are converging faster than RL deconvolution! If you haven't done so yet, run `pyflakes` over the files you have added. It will tell you about unused imports or variables.

Aretno committed:
Remove white spaces and fix alignments

Aretno committed:
Change string to Enum in deconvolution functions.

Aretno committed:
Ask for dimensional coherence and forbid n_dim>2 deconvolution.

Aretno committed:
Change array copy into asarray conversion

Aretno committed:
Use intermediate variables to clean the code

Aretno committed:
Refactoring deconvolve_hits function

Aretno committed:
Refactoring apply_deconvolution using pandas `apply` (rejected)

Although both the initial and the `apply` approach are, in nature, the
same, the `apply` approach has proven to be almost a factor two slower
than the initial implementation. This is due to the fact that apply is
called twice over the first group in ICs current version of
pandas (0.23).

This is solved in later versions so we could recover this
implementation once pandas is updated on IC.

cProfiles:
- Initial implementation:
https://github.com/Aretno/IC_profiles/blob/master/PR666_apply_deconvolution_initial_version.prof
- `apply` implementation (current pandas, 0.23):
https://github.com/Aretno/IC_profiles/blob/master/PR666_apply_deconvolution_commit_c90e327b_current_pandas.prof
- `apply` implementation (latest pandas, 1.0):
https://github.com/Aretno/IC_profiles/blob/master/PR666_apply_deconvolution_commit_c90e327b_latest_pandas.prof

Aretno committed:
Assure int in linspace

Aretno committed:
Use .loc to modify existing column in dataframe

Aretno committed:
Hardcode writer parameters

Aretno committed:
Use item in fl.map

Input-output of the functions is the same as the object is modified,
no need to create new objects for the output.

Aretno committed:
Refactor initial `apply_deconvolution`

After the `apply` approach was observed to be too slow, this commit
rolls back to the initial approach of the function with the small
changes suggested by reviewer.

Aretno committed:
Use `np.isclose` for assertion

Aretno committed:
Added fixture for city configuration in tests

Aretno committed:
Reduce if statement noise and improve empty dataframe return

Aretno committed:
Change enumerate into zip

Aretno committed:
Remove 3D option (following removal in beersheba)

Aretno committed:
Clarify differences between skimage and our RL function

Aretno committed:
Changed custom composite to hypothesis.data_frames

Aretno committed:
Removed random choice from hypothesis test

Aretno committed:
Remove unnecessary loop

Aretno committed:
Shorten long lines

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,355 @@
+    df['Z']     = [hits.Z    .unique()[0]] * len(ene) if n_dim == 2 else pos[2][sel_deconv]
+    df['X']     = pos[0][sel_deconv]
+    df['Y']     = pos[1][sel_deconv]
+    df['E']     = ene
```
```suggestion
    df['E']     = deconv_e[sel_deconv]
    df['event'] = hits.event.unique()[0]
    df['npeak'] = hits.npeak.unique()[0]
    df['Z']     = hits.Z    .unique()[0] if n_dim == 2 else pos[2][sel_deconv]
    df['X']     = pos[0][sel_deconv]
    df['Y']     = pos[1][sel_deconv]
```

Aretno committed:
Update reader to handle latest esmeralda version and data structures

Aretno committed:
Added filter for empty hdst after `drop_isolated`

After dropping isolated sensors, there is a tiny chance that the hdst
is empty if the slice only has, for example, two sensors which are
away from each other. These events are of no interest and dropped.

Aretno committed:
Update tests to the new reader.

Aretno committed:
Pyflakes magic

Aretno committed:
Cleaner dataframe value assignment

Aretno committed:
Add protection to DeconvolutionMode parameter in `deconvolve_signal`

Aretno committed:
Substitute asarray by array in non-recommendable cases.

Aretno committed:
Remove not needed arguments in `deconv_writer`

Aretno committed:
Added missing functions descriptions and comments

Aretno committed:
Change 'hdst' naming into 'cdst'

Beersheba uses esmeralda output, therefore hdst naming was
incorrect. cdst should be the naming used.

Aretno committed:
Remove one unnecessary groupby.

Groupby was redundatly done twice when could be done just one inside
the loop, this should slightly reduce the running time.

Aretno committed:
Change order of `deconvolve_signal` parameters

n_iterations_g should be last as it a configuration specific parameter

Aretno committed:
Improve readibility

Aretno committed:
Remove hard-coded number for unit change and use IC built-in

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,429 @@
+
+    df  = pd.DataFrame(columns=['event', 'npeak', 'X', 'Y', 'Z', 'E'])
+
+    if   cut_type is CutType.abs:
+        sel_deconv = deconv_e > e_cut
+    elif cut_type is CutType.rel:
+        sel_deconv = deconv_e / deconv_e.max() > e_cut
```
missing `else`

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,429 @@
+    '''
+    cut = cut_and_redistribute_df(f"Q > {q_cut}", redist_var)
+
+    def cut_over_Q(df):
```
Specify which dataframe you expect here (an inline comment is good enough)

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,429 @@
+    """
+    drop = drop_isolated_sensors(distance, redist_var)
+
+    def drop_isolated(df):
```
Specify which dataframe you expect here (an inline comment is good enough)

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,429 @@
+        if len(df) > 0:
+            return True
+        else:
+            return False
```
```suggestion
    def filter_events(df) -> bool:
        return len(df) > 0
```

gonzaponte replied:
Also, the name of this function should be generic, something like `check_nonempty_dataframe` or something like that. Then the component that uses it can have a more case-specific name.

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,429 @@
+    return drop_isolated
+
+
+def events_filter_no_hits() -> Callable:
```
this function just returns a function that is defined inside. Why not define the inner function outside and that's it?

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,429 @@
+    return write_deconv
+
+
+def summary_writer(h5out, compression='ZLIB4'):
```
is this the same summary as in `esmeralda`? If so the writer should already exist and be reused for this case.

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,429 @@
+
+    for p in ['sample_width', 'bin_size', 'diffusion']:
+        if len(deconv_params[p]) != deconv_params['n_dim']:
+            raise ValueError         ("Parameter {p} dimensions do not match n-dim parameter")
```
```suggestion
            raise ValueError         ("Parameter {p} dimensions do not match n_dim parameter")
```

gonzaponte commented on invisible_cities/cities/beersheba.py:
```diff
@@ -0,0 +1,429 @@
+                                          write_summary   ,
+                                          write_event_info))            ,
+                    result = dict(events_in  = event_count_in .future,
+                                  events_out = event_count_out.future))
```
You may want to add `events_passes_no_hits.future`

gonzaponte commented on invisible_cities/cities/beersheba_test.py:
```diff
@@ -0,0 +1,73 @@
+        assert "DECO/Events"  in h5out.root
+        assert "Run"          in h5out.root
+        assert "Run/events"   in h5out.root
+        assert "Run/runInfo"  in h5out.root
```
And the summary?

gonzaponte commented on invisible_cities/cities/beersheba_test.py:
```diff
@@ -0,0 +1,73 @@
+
+    tables = ( "MC/extents"  , "MC/hits"     , "MC/particles" , "MC/generators",
+               "DECO/Events" ,
+               "Run/events"  , "Run/runInfo" )
```
shouldn't you also check the summary table?

gonzaponte commented on invisible_cities/cities/components.py:
```diff
@@ -283,6 +283,36 @@ def pmap_from_files(paths):
             # single event) at a time.
 
+def cdst_from_files(paths: List[str]) -> Iterator[Dict[str,Union[pd.DataFrame, MCInfo, int, float]]]:
+    """Reader of the files, yields HitsCollection, pandas DataFrame with kdst info, mc_info, run_number, event_number and timestamp"""
```
Split docstring in multiple lines for readability

gonzaponte replied:
yields *corrected* hits?

gonzaponte commented on invisible_cities/conftest.py:
```diff
@@ -732,3 +739,32 @@ def dbnext100():
 def db(request):
     return request.param
+
+@pytest.fixture(scope='function')
```
why this scope?

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -0,0 +1,317 @@
+            allbins = [np.linspace(*rang, np.ceil(np.diff(rang)/sw)+1) for rang ,   sw in zip(ranges[:2], sample_width)]
+            Hs, edges = np.histogramdd(data, bins=allbins, normed=False, weights=weight)
+        else:
+            Hs, edges = np.histogramdd(data, bins=nbin   , normed=False, weights=weight, range=ranges)
```
Following the principle of not allowing values that are not a member of `InterpolationMethod`, I propose:
```
if inter_method in (InterpolationMethod.linear, InterpolationMethod.cubic):
    ...
elif inter_method is InterpolationMethod.none:
    ...
else:
    raise ValueError(...)
```

gonzaponte commented on invisible_cities/reco/deconv_functions.py:
```diff
@@ -0,0 +1,317 @@
+    H1         : Interpolated distribution weights.
+    new_points : Interpolated coordinates.
+    """
+    new_points   = np.meshgrid(*(shift_to_bin_centers(np.linspace(np.min(edge), np.max(edge), n + 1)) for n, edge in zip(nbin, edges)), indexing='ij')
```
try to break up this line a bit. you can even define the inner tuple in a different line

gonzaponte commented on invisible_cities/reco/deconv_functions_test.py:
```diff
@@ -0,0 +1,172 @@
+
+@given(data_frames(columns=[column('A', dtype=float, elements=floats(1, 1e3, allow_nan=False, allow_infinity=False)),
+                            column('B', dtype=float, elements=floats(1, 1e3, allow_nan=False, allow_infinity=False)),
+                            column('C', dtype=float, elements=floats(1, 1e3, allow_nan=False, allow_infinity=False))],
```
```suggestion
@given(data_frames(columns=[column('A', dtype=float, elements=floats(1, 1e3)),
                            column('B', dtype=float, elements=floats(1, 1e3)),
                            column('C', dtype=float, elements=floats(1, 1e3))],
```

> If min_value or max_value is not None, it is an error to enable allow_nan.
> If both min_value and max_value are not None, it is an error to enable allow_infinity.

gonzaponte commented on invisible_cities/reco/deconv_functions_test.py:
```diff
@@ -0,0 +1,172 @@
+    points = np.meshgrid(np.linspace(0, 1, 6), np.linspace(0, 1, 6)) # Coordinates where g is known
+    points = (points[0].flatten(), points[1].flatten())
+    values = g.pdf(list(zip(points[0], points[1]))) # Value of g at the known coordinates.
+    n_interpolation = 12 #How many points to interpolate
```
```suggestion
    grid_x, grid_y  = np.mgrid[0:1:12j, 0:1:12j] # Grid for interpolation
    points          = np.meshgrid(np.linspace(0, 1, 6), np.linspace(0, 1, 6)) # Coordinates where g is known
    points          = (points[0].flatten(), points[1].flatten())
    values          = g.pdf(list(zip(points[0], points[1]))) # Value of g at the known coordinates.
    n_interpolation = 12 # How many points to interpolate
```

gonzaponte commented on invisible_cities/reco/deconv_functions_test.py:
```diff
@@ -0,0 +1,172 @@
+                                           InterpolationMethod.cubic)
+    inter_charge      = out_interpolation[0].flatten()
+    inter_position    = out_interpolation[1]
+    ref_position = shift_to_bin_centers(np.linspace(-0.05, 1.05, n_interpolation + 1))
```
```suggestion
    inter_charge      = out_interpolation[0].flatten()
    inter_position    = out_interpolation[1]
    ref_position      = shift_to_bin_centers(np.linspace(-0.05, 1.05, n_interpolation + 1))
```

gonzaponte commented on invisible_cities/reco/deconv_functions_test.py:
```diff
@@ -0,0 +1,172 @@
+    h = h.groupby(['X', 'Y']).Q.sum().reset_index()
+    h = h[h.Q>40]
+    interpolator = deconvolution_input([10., 10.], [1., 1.], InterpolationMethod.cubic)
+    inter = interpolator((h.X, h.Y), h.Q)
```
```suggestion
    ref_interpolation = np.load (data_hdst_deconvolved)
    hdst              = load_dst(data_hdst, 'RECO', 'Events')
    
    h = hdst[(hdst.event == 3021916) & (hdst.npeak == 0)]
    z = h.Z.mean()
    h = h.groupby(['X', 'Y']).Q.sum().reset_index()
    h = h[h.Q > 40]
    
    interpolator = deconvolution_input([10., 10.], [1., 1.], InterpolationMethod.cubic)
    inter = interpolator((h.X, h.Y), h.Q)
```

gonzaponte commented on invisible_cities/reco/deconv_functions_test.py:
```diff
@@ -0,0 +1,172 @@
+    x, y = np.linspace(-49.5, 49.5, 100), np.linspace(-49.5, 49.5, 100)
+    xx, yy = np.meshgrid(x, y)
+    xx, yy = xx.flatten(), yy.flatten()
+    psf = {}
```
```suggestion
    x , y  = np.linspace(-49.5, 49.5, 100), np.linspace(-49.5, 49.5, 100)
    xx, yy = np.meshgrid(x, y)
    xx, yy = xx.flatten(), yy.flatten()
    
    psf = {}
```

gonzaponte commented on invisible_cities/reco/deconv_functions_test.py:
```diff
@@ -0,0 +1,172 @@
+    psf['xr']     = xx
+    psf['yr']     = yy
+    psf['zr']     = [z] * len(xx)
+    psf   = pd.DataFrame(psf)
```
```suggestion
    psf['zr']     = [z] * len(xx)
    psf           = pd.DataFrame(psf)
```

gonzaponte reviewed (commented):
Last round. Mostly cosmetical.

Aretno committed:
Add test for separate 'deconv_mode'

Aretno committed:
Fix bug in separate 'deconv_mode'

Aretno committed:
Raise error if CutType incorrect or not supported

Aretno committed:
Added missing argument description

Aretno committed:
Change `filter_events` name to a more detailed one

Aretno committed:
Simplify `check_nonempty_dataframe`

Aretno committed:
Define `check_nonempty_dataframe` as a simple function

Before it was a function calling another function which was not
optimal and did not make much sense.

Aretno committed:
Use Esmeralda `summary_writer`

Aretno committed:
Correct small typo in print

Aretno committed:
Added 'Summary' table check to tests

Aretno committed:
Clarification regarding scope declaration

Two arguments of the function are pop out from the configuration
dictionary when beersheba is executed. In a 'session' scope the config
dictionary would then be modified permanently for the coming tests
causing them to fail. Using a 'function' scope redeclares the
configuration dictionary solving this.

Aretno committed:
Improve readability of Docstring description

Aretno committed:
Add protection agains non valid interpolation methods

Aretno committed:
Improve readability

Aretno committed:
Remove unnecessary arguments test declarations

Aretno committed:
Improve spacing in `deconv_functions_test`

Aretno committed:
Add event pass filter info

gonzaponte reviewed (approved):
This PR introduces a new city that implements a new reconstruction technique for the experiment. The code is long and dense, but it is well tested. There are some functions that may be of interest outside this new city, hence there may be another PR in the future addressing that.

@Aretno has included a couple of notebooks with explanations and time profiling:
> Auxiliar notebook for the PR showing how each function works:
> https://github.com/Aretno/Notebooks/blob/master/Beersheba%20auxiliar%20Notebook.ipynb
>
> Profile (cProfile) of the city over 100 events:
> https://github.com/Aretno/Notebooks/blob/master/outprof.prof

It has been a long way to the approval of this PR. Great job!

Aretno committed:
Add test to check dimensional coherence

Aretno committed:
Add protection for enum types in `deconvolve_signal`

Aretno committed:
Add test to check enum protection in `deconvolve_signal`

Aretno committed:
Add test for CutType protection in `create_deconvolution_df`

Aretno committed:
Add protection for InterpolationMethod in `deconvolution_input`

Aretno committed:
Add test to check InterpolationMethod protection

Aretno committed:
Fix missing f-string

Aretno committed:
Add expandvars to psf path config parameter

Aretno committed:
Use IC standard function to test cities on empty files

Aretno committed:
Copy some cut dataframes to avoid `SettingWithCopyWarning`

Aretno committed:
Add beersheba to `test_command_line_run`

Aretno committed:
Add test to check expansion of `psf_fname`

gonzaponte commented on invisible_cities/cities/beersheba_test.py:
```diff
@@ -103,3 +107,23 @@ def test_beersheba_param_dim(ICDATADIR, deconvolution_config, ndim):
+    conf_dict['energy_type' ] = HitEnergy          (conf_dict['energy_type' ])
+    conf_dict['inter_method'] = InterpolationMethod(conf_dict['inter_method'])
+
+    conf_dict[param_name]     = 'check'
```
```suggestion
    conf_dict[param_name]     = param_name
```
It is particularly interesting to check that it doesn't work even if you try to use the name of the parameter, as it is the most-likely error.

gonzaponte commented on invisible_cities/cities/beersheba_test.py:
```diff
@@ -29,6 +29,16 @@ def test_create_deconvolution_df(ICDATADIR):
+        create_deconvolution_df(true_dst, true_dst.E.values,
+                                (true_dst.X.values, true_dst.Y.values, true_dst.Z.values),
+                                'check', ecut, 3)
+
```
```suggestion
@mark.parametrize("cut_type", CutType.__members__)
def test_create_deconvolution_df_cuttype(ICDATADIR, cut_type):
    true_in  = os.path.join(ICDATADIR, "exact_Kr_deconvolution_with_MC.h5")
    true_dst = dio.load_dst(true_in, 'DECO', 'Events')
    ecut     = 1e-2

    with raises(ValueError):
        create_deconvolution_df(true_dst, true_dst.E.values,
                                (true_dst.X.values, true_dst.Y.values, true_dst.Z.values),
                                cut_type, ecut, 3)

```

gonzaponte commented on invisible_cities/reco/deconv_functions_test.py:
```diff
@@ -117,6 +118,11 @@ def test_deconvolution_input(data_hdst, data_hdst_deconvolved):
+    with raises(ValueError):
+        deconvolution_input([10., 10.], [1., 1.], 'check')
+
+
```
```suggestion
@mark.parametrize("interp_method", InterpolationMethod.__members__)
def test_deconvolution_input_interpolation_method(data_hdst, data_hdst_deconvolved, interp_method):
    with raises(ValueError):
        deconvolution_input([10., 10.], [1., 1.], interp_method)


```

gonzaponte commented on invisible_cities/reco/deconv_functions_test.py:
```diff
@@ -2,6 +2,7 @@
 import numpy  as np
 import pandas as pd
 
+from pytest                  import raises
```
```suggestion
from pytest                  import mark
from pytest                  import raises
```
In order to parametrize the test below

gonzaponte reviewed (commented):
Some small improvements on the tests. We are done after this.

Aretno committed:
Use enum member list instead of standar string for protection tests
